race-buffer/src/lib.rs

```diff
@@ -2,227 +2,37 @@
-use core::fmt;
-#[cfg(feature = "std")]
-use std::fmt;
 
 pub trait Entry: Copy + PartialEq {
-    const NIL_VAL: Self;
+    //const NIL_VAL: Self;
     fn is_prefix(&self) -> bool;
 }
 
-pub trait Snapper<E>
-where
-    E: Entry,
-{
-    fn snap_wcurs(&self) -> usize;
-    fn snap_storage(&self, index: usize) -> E;
-}
-
-pub struct SizeError();
-
-impl fmt::Debug for SizeError {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str("Storage size must be a power of 2")
-    }
-}
-
 #[inline]
 /// Returns the corresponding index in backing storage of given cursor index
-fn get_cursor_index(storage_cap: usize, cursor: usize) -> usize {
-    // Index is lowest n bits of cursor where storage_cap is 2^n
-    cursor & (storage_cap - 1)
-}
-#[repr(C)]
-/// Struct used to write to buffer
-pub struct RaceBuffer<'a, E>
-where
-    E: Entry,
-{
-    /// Writer's cursor
-    wcurs: usize,
-    /// Backing storage
-    storage: &'a mut [E],
-}
-
-impl<'a, E> RaceBuffer<'a, E>
-where
-    E: Entry,
-{
-    /// Create new RaceBuffer. Returns error if storage size is not power of 2
-    pub fn new(storage: &'a mut [E]) -> Result<RaceBuffer<'a, E>, SizeError> {
-        let exp = storage.len().trailing_zeros();
-        if storage.len() != 1usize << exp {
-            return Err(SizeError());
-        }
-
-        Ok(RaceBuffer { wcurs: 0, storage })
-    }
-
-    /// Write single entry to buffer
-    pub fn write(&mut self, data: E) {
-        let write_offset = get_cursor_index(self.storage.len(), self.wcurs);
-        if self.wcurs > 0 {
-            let before_offset = get_cursor_index(self.storage.len(), self.wcurs - 1);
-            if self.storage[before_offset] == E::NIL_VAL {
-                // Previous write left free nil value behind wcurs
-                self.storage[before_offset] = data;
-                return;
-            }
-        }
-
-        if self.storage[write_offset] != E::NIL_VAL && self.storage[write_offset].is_prefix() {
-            // Overwriting double entry, must first overwrite suffix
-            let suffix_offset = get_cursor_index(self.storage.len(), self.wcurs + 1);
-            self.storage[suffix_offset] = E::NIL_VAL;
-            self.storage[write_offset] = E::NIL_VAL;
-            self.wcurs += 2;
-        } else {
-            self.storage[write_offset] = E::NIL_VAL;
-            self.wcurs += 1;
-        }
-        self.storage[write_offset] = data;
-    }
-
-    /// Get value of backing storage corresponding to given cursor
-    pub fn read_storage(&self, curs: usize) -> E {
-        self.storage[get_cursor_index(self.storage.len(), curs)]
-    }
-
-    /// Get current value of write cursor
-    pub fn read_wcurs(&self) -> usize {
-        self.wcurs
+fn get_cursor_index(storage_cap: usize, cursor: usize, use_base_2_indexing: bool) -> usize {
+    if use_base_2_indexing {
+        // Index is lowest n bits of cursor where storage_cap is 2^n
+        cursor & (storage_cap - 1)
```

Comment:
If your not already testing for this elsewhere, might be nice to have an extra debug check here:
`debug_assert!(is_power_of_two(storage_cap));`

Reply:
I'll add an explicit test for this